services/m3aggregator/config/runtime.go

```diff
@@ -0,0 +1,189 @@
+		logger.Errorf("unable to determine per-shard write new metric limit: %v", err)
+	}
+	logger.Infof("current write new metric limit per shard per second is: %d", newMetricPerShardLimit)
+
+	runtimeOpts := runtime.NewOptions().
+		SetWriteNewMetricNoLimitWarmupDuration(c.WriteNewMetricNoLimitWarmupDuration).
+		SetWriteValuesPerMetricLimitPerSecond(valueLimit).
+		SetWriteNewMetricLimitPerShardPerSecond(newMetricPerShardLimit)
+	runtimeOptsManager.SetRuntimeOptions(runtimeOpts)
+
+	valueLimitWatch, err := store.Watch(valueLimitKey)
+	if err != nil {
+		logger.Errorf("unable to watch per-metric write value limit: %v", err)
+	} else {
+		valueLimitCh = valueLimitWatch.C()
+	}
+	newMetricLimitWatch, err := store.Watch(newMetricClusterLimitKey)
+	if err != nil {
+		logger.Errorf("unable to watch cluster-wide write new metric limit: %v", err)
+	} else {
+		newMetricLimitCh = newMetricLimitWatch.C()
+	}
+	// If watch creation failed for both, we return immediately.
+	if valueLimitCh == nil && newMetricLimitCh == nil {
+		return
+	}
+
+	go func() {
+		for {
+			select {
+			case <-valueLimitCh:
+				valueLimitVal := valueLimitWatch.Get()
+				newValueLimit, err := kvutil.Int64FromValue(valueLimitVal, valueLimitKey, defaultValueLimit, nil)
+				if err != nil {
+					logger.Errorf("unable to determine per-metric write value limit: %v", err)
+					continue
+				}
+				currValueLimit := runtimeOpts.WriteValuesPerMetricLimitPerSecond()
+				if newValueLimit == currValueLimit {
+					logger.Infof("per-metric write value limit %d is unchanged, skipping", newValueLimit)
+					continue
+				}
+				logger.Infof("updating per-metric write value limit from %d to %d", currValueLimit, newValueLimit)
+				runtimeOpts = runtimeOpts.SetWriteValuesPerMetricLimitPerSecond(newValueLimit)
+				runtimeOptsManager.SetRuntimeOptions(runtimeOpts)
+			case <-newMetricLimitCh:
+				newMetricLimitVal := newMetricLimitWatch.Get()
+				var newNewMetricPerShardLimit int64
+				newNewMetricClusterLimit, err := kvutil.Int64FromValue(newMetricLimitVal, newMetricClusterLimitKey, defaultNewMetricClusterLimit, nil)
+				if err == nil {
+					newNewMetricPerShardLimit, err = clusterLimitToPerShardLimit(newNewMetricClusterLimit, placementManager)
+				}
+				if err != nil {
+					logger.Errorf("unable to determine per-shard new metric limit: %v", err)
+					continue
+				}
+				currNewMetricPerShardLimit := runtimeOpts.WriteNewMetricLimitPerShardPerSecond()
+				if newNewMetricPerShardLimit == currNewMetricPerShardLimit {
+					logger.Infof("per-shard write new metric limit %d is unchanged, skipping", newNewMetricPerShardLimit)
+					continue
+				}
+				logger.Infof("updating per-shard write new metric limit from %d to %d", currNewMetricPerShardLimit, newNewMetricPerShardLimit)
+				runtimeOpts = runtimeOpts.SetWriteNewMetricLimitPerShardPerSecond(newNewMetricPerShardLimit)
+				runtimeOptsManager.SetRuntimeOptions(runtimeOpts)
+			}
+		}
+	}()
+}
+
+func clusterLimitToPerShardLimit(
```

Comment:
Is it easier to just set the limit per shard rather than per cluster?

Reply:
It's easier to write code that way, but harder to reason about for the person who needs to set it. Say you want to set the limit, you can say I want to limit the new metric rate for the agg tier to be capped at 1M/s, and you just set it. However, if the limit is per shard, you'll need to figure out how many shards there are, what the replication factor is, and do math to figure that out, which is worse operationally.

Reply:
I thought this value is mainly to rate limit the map insertion in each agg box? if it's set per cluster, then you need to use x/8k*num_shard_per_instance to know how many insertions is happening on each m3agg box?